engine/src/test/java/io/camunda/zeebe/engine/processing/message/MessageStartEventSubscriptionTest.java

```diff
@@ -164,6 +165,36 @@ public void shouldHaveSameSubscriptionKey() {
             tuple(MessageStartEventSubscriptionIntent.DELETED, subscriptionKey));
   }
 
+  @Test // see #4099
+  public void
+      shouldResolveCorrelationKeyDefinedInMessageWhenOpeningSubscriptionForEventSubprocess() {
+    final var process =
+        Bpmn.createExecutableProcess("process")
+            .eventSubProcess(
+                "subprocess",
+                s ->
+                    s.startEvent()
+                        .message(
+                            m ->
+                                m.name("event_message")
+                                    .zeebeCorrelationKeyExpression("correlation_key"))
+                        .endEvent())
+            .startEvent()
+            .message("start_message")
+            .endEvent()
+            .done();
+
+    engine.deployment().withXmlResource(process).deploy();
+    engine
+        .message()
+        .withName("start_message")
+        .withCorrelationKey("")
+        .withVariables(Map.of("correlation_key", "key"))
+        .publish();
+
+    assertThat(RecordingExporter.incidentRecords().collect(Collectors.toList())).hasSize(0);
```

Comment:
This assertion will be slow because we can't limit the records when you expect no records matching the requirements. In this case, you expect no incidents, but we can't limit to `0` because then it would ignore if 1 or more incidents were actually written. 

🔧  Instead, we can assert that which we do expect on the log: the message subscription created event with the correctly evaluated correlation key

Something like:
```java
assertThat(
  RecordingExporter.messageSubscriptionRecords(MessageSubscriptionIntent.CREATED)
    .withMessageName("event_message")
    .withCorrelationKey("key")
    .findAny()
).isPresent();
``` 

🤔  Although this improves this specific test case, I've encountered others cases where limiting was not possible AND where the test needed to assert that a specific 'record does not appear on the log'. If you happen to have ideas on how to improve this, then I'd love to hear your thoughts on it.